src/librustdoc/doctest/rust.rs

```diff
@@ -126,23 +128,46 @@ impl HirCollector<'_> {
             return;
         }
 
+        let source_map = self.tcx.sess.source_map();
         // Try collecting `#[doc(test(attr(...)))]`
         let old_global_crate_attrs_len = self.collector.global_crate_attrs.len();
-        for doc_test_attrs in ast_attrs
-            .iter()
-            .filter(|a| a.has_name(sym::doc))
-            .flat_map(|a| a.meta_item_list().unwrap_or_default())
-            .filter(|a| a.has_name(sym::test))
-        {
-            let Some(doc_test_attrs) = doc_test_attrs.meta_item_list() else { continue };
-            for attr in doc_test_attrs
-                .iter()
-                .filter(|a| a.has_name(sym::attr))
-                .flat_map(|a| a.meta_item_list().unwrap_or_default())
-                .map(pprust::meta_list_item_to_string)
-            {
-                // Add the additional attributes to the global_crate_attrs vector
-                self.collector.global_crate_attrs.push(attr);
+        for attr in ast_attrs {
+            let Attribute::Parsed(AttributeKind::Doc(d)) = attr else { continue };
+            for attr_span in &d.test_attrs {
+                // FIXME: This is ugly, remove when `test_attrs` has been ported to new attribute API.
+                if let Ok(snippet) = source_map.span_to_snippet(*attr_span)
+                    && let Ok(stream) = TokenStream::from_str(&snippet)
+                {
+                    let mut iter = stream.into_iter().peekable();
+                    while let Some(token) = iter.next() {
+                        if let TokenTree::Ident(i) = token {
+                            let i = i.to_string();
+                            let peek = iter.peek();
+                            match peek {
+                                Some(TokenTree::Group(g)) => {
```

Comment:
I don't fully understand the cases in this match here, could you add some clarifying comments?